packages/odyssey-react-mui/src/test-selectors/querySelector.ts

```diff
@@ -11,188 +11,238 @@
-  }
-
-  return interpolatedString;
-};
-
-const getByQuerySelector = ({
-  canvas,
-  method,
-  options,
-  role,
-  text,
-}: {
-  canvas: BoundFunctions<typeof queries>;
-  method: "ByRole" | "ByLabelText" | "ByPlaceholderText" | "ByText";
-  options?: ByRoleOptions;
-  role?: Parameters<GetByRole>[1];
-  text?: Parameters<GetByText>[1];
-}) => {
-  if (method === "ByRole") {
-    return canvas.getByRole(
-      // TODO: These should eventually reference `query` as the function identifier.
-      role!,
-      options,
-    );
-  } else if (method === "ByLabelText") {
-    return canvas.getByLabelText(
-      // These should eventually reference `query` as the function identifier.
-      text!, // TODO: Use TypeScript `Infer` to ensure `label` is required when it's `ByLabelText`.
-      options,
-    );
-  } else if (method === "ByPlaceholderText") {
-    return canvas.getByPlaceholderText(
-      // These should eventually reference `query` as the function identifier.
-      text!, // TODO: Use TypeScript `Infer` to ensure `label` is required when it's `ByLabelText`.
-      options,
-    );
-  } else if (method === "ByText") {
-    return canvas.getByText(
-      // These should eventually reference `query` as the function identifier.
-      text!, // TODO: Use TypeScript `Infer` to ensure `description` is required when it's `ByLabelText`.
-      options,
-    );
-  }
-
-  return null;
-};
-
-export const querySelector = <TestSelectors extends FeatureTestSelector>({
-  canvas,
-  templateArgs: templateArgsProp,
-  testSelectors,
-}: {
+import { getComputedAccessibleText } from "./getComputedAccessibleText";
+import {
+  getByRoleQuerySelector,
+  getByTextQuerySelector,
+  type QueryMethod,
+} from "./getByQuerySelector";
+import { getControlledElement } from "./linkedHtmlSelectors";
+import { ElementError } from "./sanityChecks";
+
+export type QuerySelectorOptions<
+  LocalFeatureTestSelector extends FeatureTestSelector,
+> = LocalFeatureTestSelector extends TestSelector
+  ? Record<
+      keyof LocalFeatureTestSelector["selector"]["options"],
+      string | RegExp
+    >
+  : Record<string, string>;
+
+export type InnerQuerySelectorProps<
+  LocalFeatureTestSelector extends FeatureTestSelector,
+  LocalQueryMethod extends QueryMethod,
+> = {
   /**
-   * Testing Library canvas. This is usually `screen`, but Storybook uses `within(canvas)`.
+   * Testing Library method used to query elements.
    */
-  canvas: BoundFunctions<typeof queries>;
-  templateArgs?: TestSelectors extends TestSelector
-    ? Record<
-        TestSelectors["selector"]["templateVariableNames"][number],
+  queryMethod?: LocalQueryMethod;
+} & (LocalFeatureTestSelector extends {
+  selector: {
+    role: infer Role;
+  };
+}
+  ? Role extends AriaRole[]
+    ? {
+        /**
+         * Role is used when you have an optional `role`; otherwise, it'd baked into the metadata.
+         */
+        role: Role[number];
+      }
+    : object
+  : object) &
```

Comment:
Any way we can use something more specific than `object` here?

Reply:
I would've preferred `{}`, but Odyssey's ESLint rules don't like it, and I dislike using `// eslint-disable-next-line @typescript-eslint/ban-types`. `{}` is the correct type though.

What I'm saying is, either this object has `{ options: something }` or it is `&`ing `{}`.